heron/metricsmgr/src/java/org/apache/heron/metricsmgr/sink/metricscache/MetricsCacheClient.java

```diff
@@ -93,12 +93,20 @@ private void addMetricsCacheClientTasksOnWakeUp() {
     Runnable task = new Runnable() {
       @Override
       public void run() {
-        while (!publishMetricsCommunicator.isEmpty()) {
-          TopologyMaster.PublishMetrics publishMetrics = publishMetricsCommunicator.poll();
+        TopologyMaster.PublishMetrics publishMetrics;
+        synchronized (publishMetricsCommunicator) {
```

Comment:
I see that previously we didn't use `synchronized` during operations on `publishMetricsCommunicator`.
What have changed? Do we now work with `publishMetricsCommunicator` in different threads?

Reply:
Yeah. Before there were only 1 producer (metrics manager) and 1 consumer (sink client) for this communicator from two different threads.

In the change, the producer can remove entries from head too so it is competing with the consumer thread. A simpler way is to avoid adding new entries into the queue, then we don't need to synchronize the producer and the consumer. However, I feel that the recent data could be more important than the old data.